README.md
# Google reCAPTCHA API Gateway REST API authorizer

This construct provides an API Gateway REST API authorizer that uses Google's reCAPTCHA service to detect abusive
traffic and allow or deny access to the API methods.

It currently only supports v3 of the reCAPTCHA service.

## Installation

### TypeScript / JavaScript

`npm install cdk-recaptcha-authorizer`

or

`yarn add cdk-recaptcha-authorizer`

### Python

`pip install cdk-recaptcha-authorizer`

### Java

```xml
<dependency>
    <groupId>com.engineal.cdk</groupId>
    <artifactId>cdk-recaptcha-authorizer</artifactId>
</dependency>
```

### C# / .Net

`dotnet add package EngineAL.CDK.RecaptchaAuthorizer`

## Usage

See https://developers.google.com/recaptcha/docs/v3 for how to integrate reCAPTCHA in your client code. When
your client code executes the reCAPTCHA client, the returned token from the Google reCAPTCHA client library
should be included in the API Gateway request in the `X-reCAPTCHA-Token` header.

The `RecaptchaAuthorizer` can be used similarly to the `TokenAuthorizer` and `RequestAuthorizer` that are
provided by the `aws-cdk-lib/aws-apigateway` package. It accepts 3 properties:

* `scoreThreshold` (default: 0.5) is the threshold below which requests will be denied.
* `allowedActions` is a list of allowed actions. When your frontend executes the reCAPTCHA client, it provides
an action name. The reCAPTCHA authorizer will verify this action is allowed and deny a request if the action
is not in this list.
* `reCaptchaSecretKey` is a `SecretKey` that provides the reCAPTCHA secret key to the reCAPTCHA authorizer.
See [Secret Key](secret-key).

The following code attaches the reCAPTCHA authorizer to the 'GET' method of the root resource.
```typescript
const authorizer = new RecaptchaAuthorizer(this, 'reCaptchaAuthorizer', {
    allowedActions: ['my-action'],
    reCaptchaSecretKey: SecretKey.fromPlainText('my-secret-key')
});

api.root.addMethod('GET', new apigateway.MockIntegration(), {
    authorizer
});
```

Authorizers can also be passed via the `defaultMethodOptions` property within the RestApi construct or the
Method construct. Unless explicitly overridden, the specified defaults will be applied across all Methods
across the RestApi or across all Resources, depending on where the defaults were specified.

### Secret Key
The lambda requires the secret key generated by the Google reCAPTCHA Admin Console. You can provide it using
3 methods: plain text, AWS SSM Parameter Store, or AWS Secrets Manager.

#### Plain text
```typescript
SecretKey.fromPlainText('my-secret-key')
```

#### SSM Parameter Store
```typescript
const parameter = ssm.StringParameter.fromStringParameterName(this, 'TestParameter', 'test-secret-key');
SecretKey.fromSsmParameter(parameter)
```

#### Secrets Manager
```typescript
const secretArn = `arn:${Stack.of(this).partition}:secretsmanager:${Stack.of(this).region}:${Stack.of(this).account}:secret:test-secret`;
const secret = secretsmanager.Secret.fromSecretArn(this, 'TestSecret', secretArn);
SecretKey.fromSecretsManager(secret)
```

This also supports JSON secrets, and you can specify an optional field to use:.
```typescript
const secretArn = `arn:${Stack.of(this).partition}:secretsmanager:${Stack.of(this).region}:${Stack.of(this).account}:secret:test-secret`;
const secret = secretsmanager.Secret.fromSecretArn(this, 'TestSecret', secretArn);
SecretKey.fromSecretsManager(secret, 'my-secret-field')
```

### Useful commands

 * `npm run build`   compile the project
 * `npm run watch`   watch for changes and compile as needed
 * `npm run package` generates libraries for all languages
 * `npm run test`    perform the jest unit tests

## License

   Copyright 2023 Aaron Lucia

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


CHANGELOG.md
## [2.2.2](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.2.1...v2.2.2) (2023-05-01)


### Bug Fixes

* allow release workflow to publish packages ([56736c6](https://github.com/engineal/cdk-recaptcha-authorizer/commit/56736c6d0dff2a14b62ccc6c5b6b6aa61887e3cd))

## [2.2.1](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.2.0...v2.2.1) (2023-05-01)


### Bug Fixes

* **deps-dev:** bump dependencies ([#1361](https://github.com/engineal/cdk-recaptcha-authorizer/issues/1361)) ([daa734a](https://github.com/engineal/cdk-recaptcha-authorizer/commit/daa734a37601e800b2ea0f3b2fbc238ffa105e98))

# [2.2.0](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.1.4...v2.2.0) (2023-02-19)


### Features

* add logRetention prop ([b19f554](https://github.com/engineal/cdk-recaptcha-authorizer/commit/b19f554297fda9bed1f6541feeb435823a1d20d8))

## [2.1.4](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.1.3...v2.1.4) (2023-02-19)


### Bug Fixes

* update dependencies that don't break jsii ([#1308](https://github.com/engineal/cdk-recaptcha-authorizer/issues/1308)) ([f8b8e2f](https://github.com/engineal/cdk-recaptcha-authorizer/commit/f8b8e2f90eb5ddd497b45b4a4e4f637c18f708f9))

## [2.1.3](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.1.2...v2.1.3) (2022-11-03)


### Bug Fixes

* minor updates to publish new version ([c9f6aa3](https://github.com/engineal/cdk-recaptcha-authorizer/commit/c9f6aa3defc42be26bd48abd34f54b4aaa23c279))

## [2.1.2](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.1.1...v2.1.2) (2022-01-17)


### Bug Fixes

* project url ([4b21fc6](https://github.com/engineal/cdk-recaptcha-authorizer/commit/4b21fc695cfb1db037556405d3d23b38c5f8f293))

## [2.1.1](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.1.0...v2.1.1) (2022-01-17)


### Bug Fixes

* use jsii-release instead ([9bf8583](https://github.com/engineal/cdk-recaptcha-authorizer/commit/9bf8583e297830a2877f4180078a8f01e767191f))

# [2.1.0](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v2.0.0...v2.1.0) (2022-01-17)


### Bug Fixes

* jsii-publish version ([e3d4311](https://github.com/engineal/cdk-recaptcha-authorizer/commit/e3d4311dc58310f6e120a761ab1789b32292a00f))


### Features

* build with jsii ([071c06f](https://github.com/engineal/cdk-recaptcha-authorizer/commit/071c06f3cf4c37f5eda478ec8385ce190f7dd89e))

# [2.0.0](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.2.4...v2.0.0) (2022-01-06)


### Bug Fixes

* disable husky in CI environment ([e8caece](https://github.com/engineal/cdk-recaptcha-authorizer/commit/e8caeceef1cc4ddf99cdf2ad04591357747b288e))


### chore

* update to AWS CDK v2 ([#527](https://github.com/engineal/cdk-recaptcha-authorizer/issues/527)) ([0aa4279](https://github.com/engineal/cdk-recaptcha-authorizer/commit/0aa4279439fa7c19082f65a68bb56c58ff81c7a7))


### BREAKING CHANGES

* The version of the AWS CDK has been upgraded to v2.
Consumers of this library will also need to upgrade to v2.

* build: use latest CodeQL workflow

## [1.2.4](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.2.3...v1.2.4) (2021-04-11)


### Bug Fixes

* update aws-cdk ([#325](https://github.com/engineal/cdk-recaptcha-authorizer/issues/325)) ([d966325](https://github.com/engineal/cdk-recaptcha-authorizer/commit/d9663250193d487cb1ab9635c4aa54dc1f2ef357))

## [1.2.3](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.2.2...v1.2.3) (2020-08-24)


### Bug Fixes

* **deps:** update deps ([#107](https://github.com/engineal/cdk-recaptcha-authorizer/issues/107)) ([bf3d640](https://github.com/engineal/cdk-recaptcha-authorizer/commit/bf3d6404dd196657a3ac9849fe77c9eec22271b2))

## [1.2.2](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.2.1...v1.2.2) (2020-08-15)


### Bug Fixes

* **deps:** update aws-cdk ([9f259a7](https://github.com/engineal/cdk-recaptcha-authorizer/commit/9f259a7b18936a18382feba91910e57012dff3b2))

## [1.2.1](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.2.0...v1.2.1) (2020-07-24)


### Bug Fixes

* lint error in test ([fba6916](https://github.com/engineal/cdk-recaptcha-authorizer/commit/fba691674a609d65e2bad2d2489e748561ea00f6))
* **deps:** update aws-cdk ([f67d578](https://github.com/engineal/cdk-recaptcha-authorizer/commit/f67d5780d5a6df007e5cb7c4cb822a8df7ad2a73))

# [1.2.0](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.1.0...v1.2.0) (2020-07-06)


### Features

* capture AWS clients with X-Ray ([247fd45](https://github.com/engineal/cdk-recaptcha-authorizer/commit/247fd4587ba7a24ba8cf3a0ccaa0bb1adc71c84b))

# [1.1.0](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.0.3...v1.1.0) (2020-06-19)


### Features

* Configure score threshold ([#16](https://github.com/engineal/cdk-recaptcha-authorizer/issues/16)) ([94bf122](https://github.com/engineal/cdk-recaptcha-authorizer/commit/94bf122310516b6017a2ec55f0de84de780a9afd))

## [1.0.3](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.0.2...v1.0.3) (2020-06-13)


### Bug Fixes

* **release:** separate release workflow was a better idea ([bcb32a3](https://github.com/engineal/cdk-recaptcha-authorizer/commit/bcb32a3a7bf4864a1a0b403f5701d6f771d20cd5))

## [1.0.2](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.0.1...v1.0.2) (2020-06-13)


### Bug Fixes

* publish built code ([0d2cb14](https://github.com/engineal/cdk-recaptcha-authorizer/commit/0d2cb14061514b11c9944520ea09fc937bce61c3))

## [1.0.1](https://github.com/engineal/cdk-recaptcha-authorizer/compare/v1.0.0...v1.0.1) (2020-06-13)


### Bug Fixes

* enable semantic-release ([7c24c2e](https://github.com/engineal/cdk-recaptcha-authorizer/commit/7c24c2e3f19a7f801981a34246d052b8fe453c66))


test/recaptcha-authorizer.test.ts
/* eslint-disable max-lines */
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as cdk from 'aws-cdk-lib/core';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import {RecaptchaAuthorizer, SecretKey} from '../lib';
import {Template} from 'aws-cdk-lib/assertions';

test('Lambda Function Created with plain text secret key', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret')
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0.5',
                SECRET_KEY: 'secret',
                SECRET_KEY_TYPE: 'PLAIN_TEXT'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

test('Lambda Function Created with ssm secret key', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    const secretKeyParameter = ssm.StringParameter.fromStringParameterName(stack, 'TestParameter', 'test-secret-key');
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromSsmParameter(secretKeyParameter)
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0.5',
                SECRET_KEY_PARAMETER: 'test-secret-key',
                SECRET_KEY_TYPE: 'SSM_PARAMETER'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

// eslint-disable-next-line max-lines-per-function
test('Lambda Function Created with secrets manager secret key', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    const secretKeySecret = secretsmanager.Secret.fromSecretAttributes(stack, 'TestSecret', {
        secretPartialArn: `arn:${stack.partition}:secretsmanager:${stack.region}:${stack.account}:secret:test-secret`
    });
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromSecretsManager(secretKeySecret)
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0.5',
                SECRET_KEY_SECRET_ARN: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition'
                            },
                            ':secretsmanager:',
                            {
                                Ref: 'AWS::Region'
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId'
                            },
                            ':secret:test-secret'
                        ]
                    ]
                },
                SECRET_KEY_TYPE: 'SECRETS_MANAGER'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

// eslint-disable-next-line max-lines-per-function
test('Lambda Function Created with secrets manager and json field', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    const secretKeySecret = secretsmanager.Secret.fromSecretAttributes(stack, 'TestSecret', {
        secretPartialArn: `arn:${stack.partition}:secretsmanager:${stack.region}:${stack.account}:secret:test-secret`
    });
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromSecretsManager(secretKeySecret, 'test-field')
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0.5',
                SECRET_KEY_FIELD: 'test-field',
                SECRET_KEY_SECRET_ARN: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition'
                            },
                            ':secretsmanager:',
                            {
                                Ref: 'AWS::Region'
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId'
                            },
                            ':secret:test-secret'
                        ]
                    ]
                },
                SECRET_KEY_TYPE: 'SECRETS_MANAGER'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

// eslint-disable-next-line max-lines-per-function
test('Request Authorizer Created', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret')
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::ApiGateway::Authorizer', {
        AuthorizerUri: {
            'Fn::Join': [
                '',
                [
                    'arn:',
                    {
                        'Fn::Select': [
                            // eslint-disable-next-line no-magic-numbers
                            1,
                            {
                                'Fn::Split': [
                                    ':',
                                    {
                                        'Fn::GetAtt': [
                                            'TestAuthorizerfunction3F10EDDB',
                                            'Arn'
                                        ]
                                    }
                                ]
                            }
                        ]
                    },
                    ':apigateway:',
                    {
                        'Fn::Select': [
                            // eslint-disable-next-line no-magic-numbers
                            3,
                            {
                                'Fn::Split': [
                                    ':',
                                    {
                                        'Fn::GetAtt': [
                                            'TestAuthorizerfunction3F10EDDB',
                                            'Arn'
                                        ]
                                    }
                                ]
                            }
                        ]
                    },
                    ':lambda:path/2015-03-31/functions/',
                    {
                        'Fn::GetAtt': [
                            'TestAuthorizerfunction3F10EDDB',
                            'Arn'
                        ]
                    },
                    '/invocations'
                ]
            ]
        },
        IdentitySource: 'method.request.header.X-reCAPTCHA-Token',
        Type: 'REQUEST'
    });

    template.hasResourceProperties('AWS::ApiGateway::Method', {
        AuthorizationType: 'CUSTOM',
        AuthorizerId: {
            Ref: 'TestAuthorizer64D89012'
        }
    });
});

// eslint-disable-next-line max-lines-per-function
test('SSM parameter read granted', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    const secretKeyParameter = ssm.StringParameter.fromStringParameterName(stack, 'TestParameter', 'test-secret-key');
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromSsmParameter(secretKeyParameter)
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: [
                        'ssm:DescribeParameters',
                        'ssm:GetParameters',
                        'ssm:GetParameter',
                        'ssm:GetParameterHistory'
                    ],
                    Effect: 'Allow',
                    Resource: {
                        'Fn::Join': [
                            '',
                            [
                                'arn:',
                                {
                                    Ref: 'AWS::Partition'
                                },
                                ':ssm:',
                                {
                                    Ref: 'AWS::Region'
                                },
                                ':',
                                {
                                    Ref: 'AWS::AccountId'
                                },
                                ':parameter/test-secret-key'
                            ]
                        ]
                    }
                }
            ],
            Version: '2012-10-17'
        }
    });
});

// eslint-disable-next-line max-lines-per-function
test('Secrets Manager read granted', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');
    const secretKeySecret = secretsmanager.Secret.fromSecretAttributes(stack, 'TestSecret', {
        secretPartialArn: `arn:${stack.partition}:secretsmanager:${stack.region}:${stack.account}:secret:test-secret`
    });
    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromSecretsManager(secretKeySecret)
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });
    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: [
                        'secretsmanager:GetSecretValue',
                        'secretsmanager:DescribeSecret'
                    ],
                    Effect: 'Allow',
                    Resource: {
                        'Fn::Join': [
                            '',
                            [
                                'arn:',
                                {
                                    Ref: 'AWS::Partition'
                                },
                                ':secretsmanager:',
                                {
                                    Ref: 'AWS::Region'
                                },
                                ':',
                                {
                                    Ref: 'AWS::AccountId'
                                },
                                ':secret:test-secret-??????'
                            ]
                        ]
                    }
                }
            ],
            Version: '2012-10-17'
        }
    });
});

test('Score threshold -3 out of bounds', () => {
    const stack = new cdk.Stack();

    // THEN
    expect(() => new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret'),
        scoreThreshold: -3
    })).toThrowError('scoreThreshold must be between 0.0 and 1.0');
});

test('Score threshold 2 out of bounds', () => {
    const stack = new cdk.Stack();

    // THEN
    expect(() => new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret'),
        scoreThreshold: 2
    })).toThrowError('scoreThreshold must be between 0.0 and 1.0');
});

test('Score threshold 0.0 within bounds', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');

    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret'),
        scoreThreshold: 0.0
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });

    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0',
                SECRET_KEY: 'secret',
                SECRET_KEY_TYPE: 'PLAIN_TEXT'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

test('Score threshold 1.0 within bounds', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');

    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret'),
        scoreThreshold: 1.0
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });

    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '1',
                SECRET_KEY: 'secret',
                SECRET_KEY_TYPE: 'PLAIN_TEXT'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});

test('Score threshold 0.7 within bounds', () => {
    const stack = new cdk.Stack();
    const api = new apigateway.RestApi(stack, 'TestAPI');

    // WHEN
    const authorizer = new RecaptchaAuthorizer(stack, 'TestAuthorizer', {
        allowedActions: ['test-action'],
        reCaptchaSecretKey: SecretKey.fromPlainText('secret'),
        scoreThreshold: 0.7
    });

    api.root.addMethod('GET', new apigateway.MockIntegration(), {
        authorizer
    });

    // THEN
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
        Environment: {
            Variables: {
                ALLOWED_ACTIONS: '["test-action"]',
                AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
                SCORE_THRESHOLD: '0.7',
                SECRET_KEY: 'secret',
                SECRET_KEY_TYPE: 'PLAIN_TEXT'
            }
        },
        Handler: 'index.handler',
        Runtime: {
            'Fn::FindInMap': ['DefaultCrNodeVersionMap', {Ref: 'AWS::Region'}, 'value']
        }
    });
});


lib/recaptcha-authorizer.function.ts
/* eslint-disable no-process-env,no-console */
import * as AWSXRay from 'aws-xray-sdk-core';
import * as http from 'http';
import * as https from 'https';
import {APIGatewayAuthorizerResult, APIGatewayRequestAuthorizerEvent} from 'aws-lambda';
import {SSM} from '@aws-sdk/client-ssm';
import {SecretsManager} from '@aws-sdk/client-secrets-manager';

AWSXRay.captureHTTPsGlobal(http);
AWSXRay.captureHTTPsGlobal(https);
AWSXRay.capturePromise();

import axios from 'axios';

const secretsManagerClient = AWSXRay.captureAWSv3Client(new SecretsManager({}));
const ssmClient = AWSXRay.captureAWSv3Client(new SSM({}));

/**
 * @returns {Promise<string>} the secret key value stored in SSM Parameter Store
 * @param {string} secretKeyParameter The name of the secret key parameter to fetch
 */
const getSsmParameterSecret = async (secretKeyParameter: string): Promise<string> => {
    const parameterResponse = await ssmClient.getParameter({
        Name: secretKeyParameter,
        WithDecryption: true
    });

    if (!parameterResponse.Parameter?.Value) {
        throw new Error('SSM parameter response missing parameter!');
    }

    return parameterResponse.Parameter.Value;
};

/**
 * @returns {Promise<string>} the secret key value stored in Secrets Manager
 * @param {string} secretKeySecretArn The ARN of the secret key to fetch
 * @param {string | undefined} field An optional field to parse from a JSON secret
 */
const getSecretsManagerSecret = async (secretKeySecretArn: string, field?: string): Promise<string> => {
    const secretResponse = await secretsManagerClient.getSecretValue({
        SecretId: secretKeySecretArn
    });

    if (!secretResponse.SecretString) {
        throw new Error('Secrets Manager secret response missing secret string!');
    }

    return field ? JSON.parse(secretResponse.SecretString)[field] : secretResponse.SecretString;
};

// eslint-disable-next-line init-declarations
let cachedSecret: string | undefined;

const getSecret = async (): Promise<string> => {
    const {SECRET_KEY_TYPE, SECRET_KEY, SECRET_KEY_PARAMETER, SECRET_KEY_SECRET_ARN, SECRET_KEY_FIELD} = process.env;

    switch (SECRET_KEY_TYPE) {
        case 'PLAIN_TEXT':
            if (!SECRET_KEY) {
                throw new Error('SECRET_KEY was not defined!');
            }

            return SECRET_KEY;
        case 'SSM_PARAMETER':
            if (!SECRET_KEY_PARAMETER) {
                throw new Error('SECRET_KEY_PARAMETER was not defined!');
            }
            if (!cachedSecret) {
                // eslint-disable-next-line require-atomic-updates
                cachedSecret = await getSsmParameterSecret(SECRET_KEY_PARAMETER);
            }

            return cachedSecret;
        case 'SECRETS_MANAGER':
            if (!SECRET_KEY_SECRET_ARN) {
                throw new Error('SECRET_KEY_SECRET_ARN was not defined!');
            }
            if (!cachedSecret) {
                // eslint-disable-next-line require-atomic-updates
                cachedSecret = await getSecretsManagerSecret(SECRET_KEY_SECRET_ARN, SECRET_KEY_FIELD);
            }

            return cachedSecret;
        default:
            throw new Error(`Unsupported secret key type ${SECRET_KEY_TYPE}`);
    }
};

export const resetSecret = (): void => {
    // eslint-disable-next-line no-undefined
    cachedSecret = undefined;
};

/**
 * @returns {APIGatewayAuthorizerResult} that allows or denies this request to the method
 * @param {'Allow' | 'Deny'} effect whether to allow or deny this request
 * @param {string} methodArn the ARN of the method for this request
 */
const generateAuthResponse = (effect: 'Allow' | 'Deny', methodArn: string): APIGatewayAuthorizerResult => ({
    policyDocument: {
        Statement: [{
            Action: 'execute-api:Invoke',
            Effect: effect,
            Resource: methodArn
        }],
        Version: '2012-10-17'
    },
    principalId: 'user'
});

const DEFAULT_SCORE_THRESHOLD = 0.5;

const scoreThreshold = process.env.SCORE_THRESHOLD ? Number(process.env.SCORE_THRESHOLD) : DEFAULT_SCORE_THRESHOLD;
const allowedActions = process.env.ALLOWED_ACTIONS ? JSON.parse(process.env.ALLOWED_ACTIONS) : [];

/**
 * Lambda authorizer handler
 * @param {APIGatewayRequestAuthorizerEvent} event - API Gateway Lambda Proxy Input Format
 * @returns {APIGatewayAuthorizerResult} API Gateway Lambda Proxy Output Format
 */
// eslint-disable-next-line max-statements
export const handler = async (event: APIGatewayRequestAuthorizerEvent): Promise<APIGatewayAuthorizerResult> => {
    const token = event.headers?.['x-recaptcha-token'];
    const {methodArn, requestContext: {identity: {sourceIp}}} = event;

    if (!token) {
        console.log(`DENY: IP: ${sourceIp}; ARN: ${methodArn}:`, 'X-reCAPTCHA-Token header missing');

        return generateAuthResponse('Deny', methodArn);
    }

    const response = await axios.post('https://www.google.com/recaptcha/api/siteverify', null, {
        params: {
            remoteip: sourceIp,
            response: token,
            secret: await getSecret()
        }
    });
    const {data: {success, score, action, hostname}} = response;

    if (!success) {
        console.log(
            `DENY: IP: ${sourceIp}; ARN: ${methodArn}; hostname: ${hostname}; score: ${score}; action: ${action}:`,
            `X-reCAPTCHA-Token: "${token}" not valid`
        );

        return generateAuthResponse('Deny', methodArn);
    }

    if (score < scoreThreshold) {
        console.log(
            `DENY: IP: ${sourceIp}; ARN: ${methodArn}; hostname: ${hostname}; score: ${score}; action: ${action}:`,
            `Score less than threshold: ${scoreThreshold}`
        );

        return generateAuthResponse('Deny', methodArn);
    }

    if (!allowedActions.includes(action)) {
        console.log(
            `DENY: IP: ${sourceIp}; ARN: ${methodArn}; hostname: ${hostname}; score: ${score}; action: ${action}:`,
            `Action not in allowed actions: ${allowedActions}`
        );

        return generateAuthResponse('Deny', methodArn);
    }

    console.log(`ALLOW: IP: ${sourceIp}; ARN: ${methodArn}; hostname: ${hostname}; score: ${score}; action: ${action}`);

    return generateAuthResponse('Allow', methodArn);
};
